### DISCUSSION TITLE
PyInstaller + wine

### DISCUSSION THREAD

**Discussion by nikoniko2027** [pyinstaller/pyinstaller#7769] [Help] [answered]
After updating PyInstaller, my program stopped running via wine (version 5 / Ubuntu 20.04). How do I fix this and what do I need to do?

**Comment by bwoodsend** [comment]
You'll have to narrow it down more than that. It what way is it not working? Does a simple `print("hello")` program show the error or does your application do something specific or use a specific library that doesn't work?

**Reply by nikoniko2027** [reply]
wine: Call from 0x7b02dae6 to unimplemented function api-ms-win-crt-runtime-l1-1-0.dll.fetestexcept, aborting
wine: Unimplemented function api-ms-win-crt-runtime-l1-1-0.dll.fetestexcept called at address 000000007B02DAE6 (thread 002e), starting debugger...
0032:err:winediag:nodrv_CreateWindow Application tried to create a window, but no driver could be loaded.
0032:err:winediag:nodrv_CreateWindow Make sure that your X server is running and that $DISPLAY is set correctly.
Unhandled exception: unimplemented function api-ms-win-crt-runtime-l1-1-0.d.fetestexcept called in 64-bit code (0x000000007b02dae6).
Register dump:
 rip:000000007b02dae6 rsp:00000000003114c0 rbp:0000000000311580 eflags:00000206 (   - --  I   - -P- )
 rax:0000000000311500 rbx:00000000003115b0 rcx:00000000003114e0 rdx:0000000000000010
 rsi:0000000000000000 rdi:0000000000000000  r8:0000000000000010  r9:00000000003115b0 r10:000000000692bb8c
 r11:0000000000000002 r12:0000000000000000 r13:00000000003119f0 r14:0000000000000002 r15:0000000000000000
Stack dump:
0x00000000003114c0:  00000000003114e0 00000000003115b0
0x00000000003114d0:  0000000000000010 0000000000000000
0x00000000003114e0:  0000000180000100 0000000000000000
0x00000000003114f0:  000000007b02dae6 0000000000000002
0x0000000000311500:  00007f8ecbd6a022 00007f8ecbd6a0aa
0x0000000000311510:  0000000000000000 0000000000000000
0x0000000000311520:  0000000000000000 0000000000000000
0x0000000000311530:  0000000000000000 0000000006a66a10
0x0000000000311540:  0000000000000008 00000000067a993d
0x0000000000311550:  0000000006a669d0 0000000000000507
0x0000000000311560:  0000000000e6af60 0000000000000000
0x0000000000311570:  0000000000000000 d3bd211725c17f00
Backtrace:
=>0 0x000000007b02dae6 DebugBreak+0xb6() in kernelbase (0x0000000000311580)
  1 0x00007f8ecbd69238 in api-ms-win-crt-runtime-l1-1-0.d (+0x9237) (0x00000000003115e0)
  2 0x00007f8ecbd691df in api-ms-win-crt-runtime-l1-1-0.d (+0x91de) (0x0000000000311750)
  3 0x0000000006836c9c EntryPoint+0xffffffffffffffff() in _multiarray_umath.cp311-win_amd (0x0000000000311750)
  4 0x00000000068396c8 EntryPoint+0xffffffffffffffff() in _multiarray_umath.cp311-win_amd (0x0000000000311750)
  5 0x000000000683c815 EntryPoint+0xffffffffffffffff() in _multiarray_umath.cp311-win_amd (0x0000000000311a60)
0x000000007b02dae6 DebugBreak+0xb6 in kernelbase: movq  0x00000000000000b8(%rsp),%rax
Modules:
Module  Address                                 Debug info      Name (127 modules)
PE               4d60000-         4d7e000       Deferred        _ctypes.pyd
PE               4d80000-         4d8b000       Deferred        libffi-8
PE               4d90000-         4da5000       Deferred        _bz2.pyd
PE               4db0000-         4dd9000       Deferred        _lzma.pyd
PE               50e0000-         5112000       Deferred        pyexpat.pyd
PE               5120000-         5135000       Deferred        _socket.pyd
PE               5140000-         5149000       Deferred        select.pyd
PE               5250000-         5259000       Deferred        _queue.pyd
PE               5760000-         5788000       Deferred        _ssl.pyd
PE               5790000-         5ae2000       Deferred        libcrypto-1_1
PE               5af0000-         5b9e000       Deferred        libssl-1_1
PE               5ba0000-         5bb1000       Deferred        _asyncio.pyd
PE               5cc0000-         5cce000       Deferred        _overlapped.pyd
PE               5cd0000-         5cdd000       Deferred        _feature_detector_cpp.cp311-win_amd64.pyd
PE               5de0000-         5f88000       Deferred        metrics_cpp.cp311-win_amd64.pyd
PE               5f90000-         5f9c000       Deferred        vcruntime140_1
PE               60b0000-         60f8000       Deferred        _initialize_cpp.cp311-win_amd64.pyd
PE               6100000-         6241000       Deferred        fuzz_cpp.cp311-win_amd64.pyd
PE               6690000-         696f000       Export          _multiarray_umath.cp311-win_amd64.pyd
PE               6a70000-         6a85000       Deferred        _multiarray_tests.cp311-win_amd64.pyd
ELF             7b000000-        7b0e2000       Dwarf           kernelbase<elf>
  \-PE          7b010000-        7b0e2000       \               kernelbase
ELF             7b400000-        7b68b000       Deferred        kernel32<elf>
  \-PE          7b420000-        7b68b000       \               kernel32
ELF             7bc00000-        7befb000       Deferred        ntdll<elf>
  \-PE          7bc20000-        7befb000       \               ntdll
ELF             7c000000-        7c005000       Deferred        <wine-loader>
PE             140000000-       140059000       Deferred        shopbot
PE             180000000-       1805cd000       Deferred        python311
PE             347970000-       34984a000       Deferred        libopenblas64__v0.3.23-gcc_10_3_0
ELF         7f8eca881000-    7f8eca8b7000       Deferred        api-ms-win-crt-private-l1-1-0<elf>
  \-PE      7f8eca890000-    7f8eca8b7000       \               api-ms-win-crt-private-l1-1-0
ELF         7f8eca9b7000-    7f8ecaaef000       Deferred        msvcr120<elf>
  \-PE      7f8eca9e0000-    7f8ecaaef000       \               msvcr120
ELF         7f8ecaaef000-    7f8ecab0f000       Deferred        concrt140<elf>
  \-PE      7f8ecab00000-    7f8ecab0f000       \               concrt140
ELF         7f8ecab0f000-    7f8ecac24000       Deferred        msvcp140<elf>
  \-PE      7f8ecab50000-    7f8ecac24000       \               msvcp140
ELF         7f8ecac24000-    7f8ecad37000       Deferred        crypt32<elf>
  \-PE      7f8ecac30000-    7f8ecad37000       \               crypt32
ELF         7f8ecad37000-    7f8ecad6b000       Deferred        iphlpapi<elf>
  \-PE      7f8ecad40000-    7f8ecad6b000       \               iphlpapi
ELF         7f8ecad6b000-    7f8ecad81000       Deferred        api-ms-win-crt-utility-l1-1-0<elf>
  \-PE      7f8ecad70000-    7f8ecad81000       \               api-ms-win-crt-utility-l1-1-0
ELF         7f8ecad81000-    7f8ecada7000       Deferred        imm32<elf>
  \-PE      7f8ecad90000-    7f8ecada7000       \               imm32
ELF         7f8ecada7000-    7f8ecadb0000       Deferred        libuuid.so.1
ELF         7f8ecadb0000-    7f8ecadde000       Deferred        libexpat.so.1
ELF         7f8ecadde000-    7f8ecae25000       Deferred        libfontconfig.so.1
ELF         7f8ecae25000-    7f8ecae41000       Deferred        libz.so.1
ELF         7f8ecae41000-    7f8ecae79000       Deferred        libpng16.so.16
ELF         7f8ecae79000-    7f8ecaf38000       Deferred        libfreetype.so.6
ELF         7f8ecaf38000-    7f8ecb092000       Deferred        oleaut32<elf>
  \-PE      7f8ecaf60000-    7f8ecb092000       \               oleaut32
ELF         7f8ecb092000-    7f8ecb129000       Deferred        rpcrt4<elf>
  \-PE      7f8ecb0a0000-    7f8ecb129000       \               rpcrt4
ELF         7f8ecb129000-    7f8ecb2e1000       Deferred        gdi32<elf>
  \-PE      7f8ecb140000-    7f8ecb2e1000       \               gdi32
ELF         7f8ecb2e1000-    7f8ecb57c000       Deferred        user32<elf>
  \-PE      7f8ecb300000-    7f8ecb57c000       \               user32
ELF         7f8ecb57c000-    7f8ecb6f0000       Deferred        ole32<elf>
  \-PE      7f8ecb5a0000-    7f8ecb6f0000       \               ole32
ELF         7f8ecb6f0000-    7f8ecb6fc000       Deferred        libffi.so.7
ELF         7f8ecb6fc000-    7f8ecb780000       Deferred        libgmp.so.10
ELF         7f8ecb780000-    7f8ecb7b7000       Deferred        libhogweed.so.5
ELF         7f8ecb7b7000-    7f8ecb7f1000       Deferred        libnettle.so.7
ELF         7f8ecb7f1000-    7f8ecb807000       Deferred        libtasn1.so.6
ELF         7f8ecb807000-    7f8ecb989000       Deferred        libunistring.so.2
ELF         7f8ecb989000-    7f8ecb9aa000       Deferred        libidn2.so.0
ELF         7f8ecb9aa000-    7f8ecbae0000       Deferred        libp11-kit.so.0
ELF         7f8ecbae0000-    7f8ecbcb6000       Deferred        libgnutls.so.30
ELF         7f8ecbcb6000-    7f8ecbccc000       Deferred        api-ms-win-crt-filesystem-l1-1-0<elf>
  \-PE      7f8ecbcc0000-    7f8ecbccc000       \               api-ms-win-crt-filesystem-l1-1-0
ELF         7f8ecbccc000-    7f8ecbce2000       Deferred        api-ms-win-crt-conio-l1-1-0<elf>
  \-PE      7f8ecbcd0000-    7f8ecbce2000       \               api-ms-win-crt-conio-l1-1-0
ELF         7f8ecbce2000-    7f8ecbcf8000       Deferred        api-ms-win-crt-heap-l1-1-0<elf>
  \-PE      7f8ecbcf0000-    7f8ecbcf8000       \               api-ms-win-crt-heap-l1-1-0
ELF         7f8ecbcf8000-    7f8ecbd0e000       Deferred        api-ms-win-crt-process-l1-1-0<elf>
  \-PE      7f8ecbd00000-    7f8ecbd0e000       \               api-ms-win-crt-process-l1-1-0
ELF         7f8ecbd0e000-    7f8ecbd24000       Deferred        api-ms-win-crt-time-l1-1-0<elf>
  \-PE      7f8ecbd10000-    7f8ecbd24000       \               api-ms-win-crt-time-l1-1-0
ELF         7f8ecbd24000-    7f8ecbd3f000       Deferred        api-ms-win-crt-math-l1-1-0<elf>
  \-PE      7f8ecbd30000-    7f8ecbd3f000       \               api-ms-win-crt-math-l1-1-0
ELF         7f8ecbd3f000-    7f8ecbd57000       Deferred        api-ms-win-crt-convert-l1-1-0<elf>
  \-PE      7f8ecbd50000-    7f8ecbd57000       \               api-ms-win-crt-convert-l1-1-0
ELF         7f8ecbd57000-    7f8ecbd6e000       Dwarf           api-ms-win-crt-runtime-l1-1-0<elf>
  \-PE      7f8ecbd60000-    7f8ecbd6e000       \               api-ms-win-crt-runtime-l1-1-0
ELF         7f8ecbd6e000-    7f8ecbd86000       Deferred        api-ms-win-crt-string-l1-1-0<elf>
  \-PE      7f8ecbd80000-    7f8ecbd86000       \               api-ms-win-crt-string-l1-1-0
ELF         7f8ecbd86000-    7f8ecbd9c000       Deferred        api-ms-win-crt-locale-l1-1-0<elf>
  \-PE      7f8ecbd90000-    7f8ecbd9c000       \               api-ms-win-crt-locale-l1-1-0
ELF         7f8ecbd9c000-    7f8ecbdb2000       Deferred        api-ms-win-crt-environment-l1-1-0<elf>
  \-PE      7f8ecbda0000-    7f8ecbdb2000       \               api-ms-win-crt-environment-l1-1-0
ELF         7f8ecbdb2000-    7f8ecbdca000       Deferred        api-ms-win-crt-stdio-l1-1-0<elf>
  \-PE      7f8ecbdc0000-    7f8ecbdca000       \               api-ms-win-crt-stdio-l1-1-0
ELF         7f8ecbdca000-    7f8ecbf0f000       Deferred        ucrtbase<elf>
  \-PE      7f8ecbe00000-    7f8ecbf0f000       \               ucrtbase
ELF         7f8ecbf0f000-    7f8ecbf26000       Deferred        vcruntime140<elf>
  \-PE      7f8ecbf20000-    7f8ecbf26000       \               vcruntime140
ELF         7f8ecbf26000-    7f8ecbf4d000       Deferred        bcrypt<elf>
  \-PE      7f8ecbf30000-    7f8ecbf4d000       \               bcrypt
ELF         7f8ecbf4d000-    7f8ecbf63000       Deferred        api-ms-win-core-path-l1-1-0<elf>
  \-PE      7f8ecbf50000-    7f8ecbf63000       \               api-ms-win-core-path-l1-1-0
ELF         7f8ecbf63000-    7f8ecbfa8000       Deferred        ws2_32<elf>
  \-PE      7f8ecbf70000-    7f8ecbfa8000       \               ws2_32
ELF         7f8ecbfa8000-    7f8ecc09f000       Deferred        msvcrt<elf>
  \-PE      7f8ecbfd0000-    7f8ecc09f000       \               msvcrt
ELF         7f8ecc09f000-    7f8ecc0b7000       Deferred        version<elf>
  \-PE      7f8ecc0b0000-    7f8ecc0b7000       \               version
ELF         7f8ecc0b7000-    7f8ecc0cd000       Deferred        api-ms-win-core-localization-l1-2-1<elf>
  \-PE      7f8ecc0c0000-    7f8ecc0cd000       \               api-ms-win-core-localization-l1-2-1
ELF         7f8ecc0cd000-    7f8ecc0e3000       Deferred        api-ms-win-core-fibers-l1-1-1<elf>
  \-PE      7f8ecc0d0000-    7f8ecc0e3000       \               api-ms-win-core-fibers-l1-1-1
ELF         7f8ecc0e3000-    7f8ecc0f9000       Deferred        api-ms-win-core-synch-l1-2-0<elf>
  \-PE      7f8ecc0f0000-    7f8ecc0f9000       \               api-ms-win-core-synch-l1-2-0
ELF         7f8ecc1f9000-    7f8ecc296000       Deferred        advapi32<elf>
  \-PE      7f8ecc210000-    7f8ecc296000       \               advapi32
ELF         7f8ecc396000-    7f8ecc3aa000       Deferred        libnss_files.so.2
ELF         7f8eccb6a000-    7f8eccb93000       Deferred        liblzma.so.5
ELF         7f8eccb93000-    7f8eccbae000       Deferred        libgcc_s.so.1
ELF         7f8eccbae000-    7f8ecccfd000       Deferred        libm.so.6
ELF         7f8ecccfd000-    7f8eccd1a000       Deferred        libunwind.so.8
ELF         7f8eccd1c000-    7f8eccd22000       Deferred        libdl.so.2
ELF         7f8eccd22000-    7f8eccf14000       Deferred        libc.so.6
ELF         7f8eccf14000-    7f8eccf37000       Deferred        libpthread.so.0
ELF         7f8eccf45000-    7f8ecd0e8000       Dwarf           libwine.so.1
ELF         7f8ecd0ea000-    7f8ecd11a000       Deferred        ld-linux-x86-64.so.2
Threads:
process  tid      prio (all id:s are in hex)
0000000e services.exe
        00000029    0
        00000026    0
        0000001f    0
        0000001c    0
        00000015    0
        00000014    0
        00000013    0
        00000010    0
        0000000f    0
00000011 plugplay.exe
        00000017    0
        00000016    0
        00000012    0
00000018 winedevice.exe
        00000021    0
        00000020    0
        0000001e    0
        0000001d    0
        00000019    0
00000024 winedevice.exe
        0000002a    0
        00000028    0
        00000027    0
        00000025    0
0000002b shopbot.exe
        0000002c    0
0000002d (D) Z:\root\ShopBot.exe
        0000002e    0 <==
00000034 explorer.exe
        00000037    0
        00000036    0
        00000035    0
System information:
    Wine build: wine-5.0 (Ubuntu 5.0-3ubuntu1)
    Platform: x86_64
    Version: Windows 7
    Host system: Linux
    Host version: 5.4.0-153-generic

**Reply by rokm** [reply]
> ```
> =>0 0x000000007b02dae6 DebugBreak+0xb6() in kernelbase (0x0000000000311580)
> 1 0x00007f8ecbd69238 in api-ms-win-crt-runtime-l1-1-0.d (+0x9237) (0x00000000003115e0)
> 2 0x00007f8ecbd691df in api-ms-win-crt-runtime-l1-1-0.d (+0x91de) (0x0000000000311750)
> 3 0x0000000006836c9c EntryPoint+0xffffffffffffffff() in _multiarray_umath.cp311-win_amd (0x0000000000311750)
> 4 0x00000000068396c8 EntryPoint+0xffffffffffffffff() in _multiarray_umath.cp311-win_amd (0x0000000000311750)
> 5 0x000000000683c815 EntryPoint+0xffffffffffffffff() in _multiarray_umath.cp311-win_amd (0x0000000000311a60)
> 0x000000007b02dae6 DebugBreak+0xb6 in kernelbase: movq 0x00000000000000b8(%rsp),%rax
> ```

`_multiarray_umath` is a `numpy` extension, so that's what's causing the crash in wine due to unimplemented CRT function. Which likely means that compared to your working build, you did not update just PyInstaller, but also `numpy`.

**Reply by nikoniko2027** [reply]
At the same time, the weight of my program increased from 8 MB to 28 MB. This is very strange

**Reply by rokm** [reply]
If you're using `numpy`, I doubt your program was 8 MB. Just `libopenblas64__v0.3.21-gcc_10_3_0.dll` they ship is 38 MB.

**Reply by rokm** [reply]
So even if we're talking about a `onefile` executable size here, a basic `import numpy` program comes out at 20.5 MB here.

**Reply by nikoniko2027** [reply]
![image](https://github.com/pyinstaller/pyinstaller/assets/38989700/4a5711bc-6090-408a-98e5-aa5d5cfdb1f6)
This is quite strange, because I don't really use numpy anywhere.

**Reply by nikoniko2027** [reply]
Amazing! I removed numpy from PyPl (all Python packages in Visual Studio 2022) and the weight of my program became 15 mb, while it also began to work on Wine

**Reply by rokm** [reply, marked as answer]
Be that as it may, something is clearly importing `numpy` in your new build, leading to that crash in wine's implementation of CRT.

So either `numpy` is required by a newer version of some package that you do use (so you have it installed now); or it is an optional dependency of a package that you use, and you didn't have it installed in your previous build environment. 

Make sure to build in an environment that has minimal set of required packages installed (so you don't end up with collected optional dependencies), or explicitly exclude packages that you have installed but know you don't need. For example, add `--exclude numpy` to your PyInstaller command (or to `excludes` list in the .spec file).

**Reply by nikoniko2027** [reply]
Thank you! I will definitely take your advice!

**Comment by nikoniko2027** [comment]
It is also worth noting that my exe file used to occupy 8 MB, and now occupies 28 MB. At the same time, it should be taken into account that I did not add anything new to my program

**Reply by bwoodsend** [reply]
Given that Windows numpy is 15MB compressed, I very much doubt that your 8MB executable contained numpy.
